pkg/exchange/coinbase/exchange.go

```diff
@@ -485,3 +475,93 @@ func (e *Exchange) DefaultFeeRates() types.ExchangeFee {
 		TakerFeeRate: fixedpoint.NewFromFloat(0.4 * 0.01),  // 40 bps = 0.4%
 	}
 }
+
+// ExchangeTradeHistoryService
+// QueryClosedOrders queries closed orders for the given symbol and time range.
+// startTime and endTime are inclusive.
+func (e *Exchange) QueryClosedOrders(ctx context.Context, symbol string, startTime, endTime time.Time, lastOrderID uint64) ([]types.Order, error) {
+	if lastOrderID > 0 {
+		log.Warning("lastOrderID is not supported for Coinbase, it will be ignored")
+	}
+	cbOrders, err := e.queryOrdersByPagination(ctx, symbol, []string{"done"})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to query closed orders for %s", symbol)
+	}
+	orders := make([]types.Order, 0)
+	for _, cbOrder := range cbOrders {
+		if cbOrder.CreatedAt.Before(startTime) || cbOrder.CreatedAt.After(endTime) {
+			continue
+		}
+		order := toGlobalOrder(&cbOrder)
+		orders = append(orders, order)
+	}
+	return orders, nil
+}
+
+func (e *Exchange) QueryTrades(ctx context.Context, symbol string, options *types.TradeQueryOptions) ([]types.Trade, error) {
+	cbTrades, err := e.queryProductTradesByPagination(
+		ctx, symbol, options,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to query trades for %s", symbol)
+	}
+	var trades []types.Trade
+	for _, cbTrade := range cbTrades {
+		trades = append(trades, toGlobalTrade(&cbTrade))
+	}
+	return trades, nil
+}
+
+func (e *Exchange) queryProductTradesByPagination(
+	ctx context.Context, symbol string, options *types.TradeQueryOptions,
+) (cbTrades api.TradeSnapshot, err error) {
+	defer func() {
+		if err == nil && options.Limit > 0 {
+			cbTrades = cbTrades[:options.Limit]
+		}
+	}()
+	req := e.client.NewGetOrderTradesRequest().Limit(PaginationLimit)
+	req.ProductID(toLocalSymbol(symbol))
+	if options.StartTime != nil {
+		req.StartDate(
+			options.StartTime.Format(time.RFC3339Nano),
+		)
+	}
+	if options.EndTime != nil {
+		req.EndDate(
+			options.EndTime.Format(time.RFC3339Nano),
+		)
+	}
+	if options.LastTradeID > 0 {
+		req.Before(options.LastTradeID)
```

Comment:
the LastTradeID is used in ascending ordering, so it needs to be used as "After"